processor/src/main/java/org/mapstruct/ap/internal/model/source/TargetReference.java

```diff
@@ -112,24 +136,31 @@ public TargetReference build() {
             boolean foundEntryMatch;
             Type resultType = method.getResultType();
 
-            // there can be 3 situations
+            // there can be 4 situations
             // 1. Return type
-            // 2. @MappingTarget, with
-            // 3. or without parameter name.
+            // 2. A reverse target reference where the source parameter name is used in the original mapping
+            // 3. @MappingTarget, with
+            // 4. or without parameter name.
             String[] targetPropertyNames = segments;
-            List<PropertyEntry> entries = getTargetEntries( resultType, targetPropertyNames );
+            List<PropertyEntry> entries = getTargetEntries( resultType, targetPropertyNames, false );
             foundEntryMatch = (entries.size() == targetPropertyNames.length);
-            if ( !foundEntryMatch && segments.length > 1 ) {
+            if ( !foundEntryMatch && segments.length > 1
+                && matchesSourceOrTargetParameter( segments[0], parameter, reverseSourceParameter, isReverse ) ) {
                 targetPropertyNames = Arrays.copyOfRange( segments, 1, segments.length );
-                entries = getTargetEntries( resultType, targetPropertyNames );
+                entries = getTargetEntries( resultType, targetPropertyNames, false );
                 foundEntryMatch = (entries.size() == targetPropertyNames.length);
             }
 
+            if ( !foundEntryMatch ) {
+                // This is called only for reporting errors
+                getTargetEntries( resultType, targetPropertyNames, true );
+            }
+
             // foundEntryMatch = isValid, errors are handled in the BeanMapping, where the error context is known
             return new TargetReference( parameter, entries, foundEntryMatch );
         }
 
-        private List<PropertyEntry> getTargetEntries(Type type, String[] entryNames) {
+        private List<PropertyEntry> getTargetEntries(Type type, String[] entryNames, boolean reportError) {
```

Comment:
Well the error analysis is always done, we need to know when to report the error. We can't do it on the first iteration as we can theoretically have the target property name in the mapping.